components/layout_2020/display_list/clip_path.rs

```diff
@@ -0,0 +1,228 @@
+                    fragment_builder.border_radius,
+                    layout_rect.size(),
+                    fragment_builder.fragment,
+                ),
+                _ => fragment_builder.border_radius,
+            },
+            layout_rect,
+            parent_scroll_node_id,
+            parent_clip_chain_id,
+            display_list,
+        )
+    }
+}
+
+fn build_simple_shape(
+    shape: BasicShape,
+    layout: LayoutRect,
+    parent_scroll_node_id: ScrollTreeNodeId,
+    parent_clip_chain_id: ClipChainId,
+    display_list: &mut super::DisplayList,
+) -> Option<ClipChainId> {
+    match shape {
+        BasicShape::Rect(rect) => {
+            let top = rect.rect.0.resolve(Length::new(layout.height()));
+            let right = rect.rect.1.resolve(Length::new(layout.width()));
+            let bottom = rect.rect.2.resolve(Length::new(layout.height()));
+            let left = rect.rect.3.resolve(Length::new(layout.width()));
+            let x = layout.min.x + left.px();
+            let y = layout.min.y + top.px();
+            let width = layout.width() - (left + right).px();
+            let height = layout.height() - (top + bottom).px();
+            let resolve = |radius: &LengthPercentage, box_size: f32| {
+                radius.percentage_relative_to(Length::new(box_size)).px()
+            };
+            let corner = |corner: &style::values::computed::BorderCornerRadius| {
+                LayoutSize::new(
+                    resolve(&corner.0.width.0, layout.size().width),
+                    resolve(&corner.0.height.0, layout.size().height),
+                )
+            };
+            let mut radii = webrender_api::BorderRadius {
+                top_left: corner(&rect.round.top_left),
+                top_right: corner(&rect.round.top_right),
+                bottom_left: corner(&rect.round.bottom_left),
+                bottom_right: corner(&rect.round.bottom_right),
+            };
+            let origin = LayoutPoint::new(x, y);
+            let size = LayoutSize::new(width, height);
+            let rect = LayoutRect::from_origin_and_size(origin, size);
+            normalize_radii(&layout, &mut radii);
+            create_rect_clip_chain(
+                radii,
+                rect,
+                parent_scroll_node_id,
+                parent_clip_chain_id,
+                display_list,
+            )
+        },
+        BasicShape::Circle(circle) => {
+            let center = match circle.position {
+                GenericPositionOrAuto::Position(position) => position,
+                GenericPositionOrAuto::Auto => Position::center(),
+            };
+            let anchor_x = center.horizontal.resolve(Length::new(layout.width()));
+            let anchor_y = center.vertical.resolve(Length::new(layout.height()));
+            let amount = LayoutSize::new(anchor_x.px(), anchor_y.px());
+            let center = layout.min.add_size(&amount);
```

Comment:
I would combine this to avoid the ambiguously-named `amount`.
```suggestion
            let center = layout.min.add_size(&LayoutSize::new(anchor_x.px(), anchor_y.px()));
```